integration/network/ipvlan/ipvlan_test.go

```diff
@@ -420,3 +463,84 @@ func ipvlanKernelSupport(t *testing.T) bool {
 
 	return ipvlanSupported
 }
+
+// TestIPVlanDNS checks whether DNS is forwarded, for combinations of l2/l3 mode,
+// with/without a parent interface, and with '--internal'. Note that, there's no
+// attempt here to give the ipvlan network external connectivity - when this test
+// supplies a parent interface, it's a dummy. External DNS lookups only work
+// because the daemon is configured to see a host resolver on a loopback
+// interface, so the external DNS lookup happens in the host's namespace. The
+// test is checking that an automatically configured dummy interface causes the
+// network to behave as if it was '--internal'. Regression test for
+// https://github.com/moby/moby/issues/47662
+func TestIPVlanDNS(t *testing.T) {
+	skip.If(t, testEnv.IsRootless, "rootless mode has different view of network")
+
+	ctx := testutil.StartSpan(baseContext, t)
+
+	net.StartDaftDNS(t, "127.0.0.1")
+
+	tmpFileName := net.WriteTempResolvConf(t, "127.0.0.1")
+	d := daemon.New(t, daemon.WithEnvVars("DOCKER_TEST_RESOLV_CONF_PATH="+tmpFileName))
+	d.StartWithBusybox(ctx, t)
+	t.Cleanup(func() { d.Stop(t) })
+	c := d.NewClientT(t)
+
+	const parentIfName = "di-dummy0"
+	n.CreateMasterDummy(ctx, t, parentIfName)
+	defer n.DeleteInterface(ctx, t, parentIfName)
+
+	const netName = "ipvlan-dns-net"
+
+	testcases := []struct {
+		name     string
+		parent   string
+		internal bool
+		expDNS   bool
+	}{
+		{
+			name:   "with parent",
+			parent: parentIfName,
+			// External DNS should be used (even though the network has no external connectivity).
+			expDNS: true,
+		},
+		{
+			name: "no parent",
+			// External DNS should not be used, equivalent to '--internal'.
+		},
+		{
+			name:     "with parent, internal",
+			parent:   parentIfName,
+			internal: true,
+			// External DNS should not be used.
+		},
+	}
+
+	for _, mode := range []string{"l2", "l3"} {
+		for _, tc := range testcases {
+			t.Run(tc.name+" "+mode, func(t *testing.T) {
```

Comment:
Please name subtests in a way which works with, rather than against, the split-on-slashes semantics of `go test -run`. This particular set of tests is a matrix with three variables: mode, has parent, is internal. If the tests were to be named as follows, it would be possible to conveniently run any slice of the test matrix by specifying which variables to hold fixed.
- `Mode=l2/HasParent=True/Internal=False`
- `Mode=l2/HasParent=False/Internal=False`
- `Mode=l2/HasParent=True/Internal=True`
- `Mode=l3/HasParent=True/Internal=False`
- `Mode=l3/HasParent=False/Internal=False`
- `Mode=l3/HasParent=True/Internal=True`

Run all L2 tests: `go test -run TestIPVLAN_DNS/Mode=l2`
Run all non-internal tests: `go test -run TestIPVLAN_DNS///Internal=False`

```suggestion
			t.Run(fmt.Sprintf("Mode=%v/HasParent=%v/Internal=%v", tc.mode, tc.parent != "", tc.internal), func(t *testing.T) {
```

Reply:
Done.